### DISCUSSION TITLE
Sensor not updating

### DISCUSSION THREAD

**Discussion by smarg120** [GuyKh/iec-custom-component#183] [Q&A] [answered]
Hi, first of all great integration! really useful...

but i might be doing something wrong as it does not seem to get updated data. Its stuck on the data from yesterday and does not update....in the logs i see it goes to IEC api which replies back with stall data....but if i login via website and call the same api (i also called it in postman to make sure),  i do get an updated data. 

anyone experienced this? thanks!

update:

i also downloaded and installed the rest api python package and tested it - it works as expected and brings updated data from the IEC API....however the HA component does not get updated and in the log i see HTTP responses from IEC with outdated data.....what the heck

**Comment by smarg120** [comment]
ok, i think i found it ...its a bug with HA component that assumes contract id is an int.....however my current contract id, after i moved to  a different company, has a leading zeros....my original contract id was X, but my current contract id is 000X.....so when switching to int, the leading zeros are lost and the request for  incorrect contract id is sent

**Reply by GuyKh** [reply]
You mean that in IEC API when sending a number and sending a (string with) leading zeroed number, you get different results?

**Reply by GuyKh** [reply]
Can you retry the py package with both the leading zerod and without and share the diff? (as well as which methods you used)

**Reply by smarg120** [reply]
ok its not that....i just tested that in postman with leading zeros vs without......the response is the same , it brings the the actual data....so its not that...i got confused since the python rest package sends request to URL with leading zeros, while HA component sends to the one without zeros....so i thought thats the issue

**Reply by smarg120** [reply]
so i am lost...i installed the integration 2 days ago and since then it shows the same consumption in HA

**Reply by smarg120** [reply]
ok i played with it a bit more and here is something weird....

HA component sends the following request - lastInvoiceDate":"2024-10-12","fromDate":"2024-10-12","resolution":3}'
the date is in the future, and the IEC respond with data for yesterday- example:

  "fromDate": "2024-10-01",
  "toDate": "2024-10-11",
  "totalConsumptionForPeriod": 481.3,
  "totalImportDateForPeriod": "2024-10-10",

if i modify the request to today date - lastInvoiceDate":"2024-10-11","fromDate":"2024-10-11","resolution":3}' - the response includes today's consumption:

"fromDate": "2024-10-01",
  "toDate": "2024-10-11",
  "totalConsumptionForPeriod": 552.579,
  "totalImportDateForPeriod": "2024-10-11",

now it looks like a bug on IEC side......what is the reason for HA component to ask for future day?

**Reply by GuyKh** [reply]
It's not a bug but a workaround for IEC data.
there were inconsistencies for future consumption (like you see now) and from my experiments tomorrow date got me the latest data. I'll take another look

**Reply by smarg120** [reply]
ok, today HA data suddenly got updated..and now it shows LARGER numbers than IEC web site. e.g. this month consumption shows 583 while web site still shows 552 - web site states the data is correct for yesterday - but yesterday HA gave me 481 for that sensor....i will keep monitoring this to try to understand whats going on

**Reply by GuyKh** [reply]
Most of that weird behavior is related to IEC (broken) API behavior. And these odd calls are workarounds for mitigation.

**Reply by smarg120** [reply]
Ok, so I think I figured it out. After watching it for a couple of days,  it keeps updating.  But here is the problem.  It brings Yesterday consumption , but it is being reported TODAY. So in statistics graph there is a 1 day shift of consumption and this totally breaks correlation with other data ( e.g. amount of time of central ac unit working etc) and makes daily consumption data unusable.  After a quick google I understand this is a HA limitation that can only record the current time of the reported state and cannot receive a timestamp for the reported value from outside. So in current form daily consumption is not really usable. Do you have any ideas how to overcome this issue? I could not find anything except suggestion to manipulate timestamps inside HA DB, which I dont like....

Thanks!

**Reply by GuyKh** [reply]
Are you sure you use the `iec:` entity in statistics view?

**Reply by smarg120** [reply]
i added sensor.latest_meter_reading.........if i add iec: entity i dont get any current readings whatsoever, the last is from september. iec entity does not report anything for october...only sensor. entities report

**Reply by GuyKh** [reply]
@smarg120 - you should use `iec:` in the statistics view.
Every hour another day would be added till it catch up.
It's designed like this not to flood IEC API

**Reply by smarg120** [reply]
i see....the issue is - i dont see anything in statistics view from  iec, besides a single total monthly report for September....the integration is installed for 4 days - so it was supposed to bring consumption for october already? but i see nothing.....is it normal? i dont see any progress for iec reports.....maybe i am checking wrong....dont know

**Reply by smarg120** [reply]
this is my energy view...the dark blue is iec:......the other one is sensor, which appears also wrong for some reason, since the total consumption for this month is already ~700 KWh ( ok got it - it shows just for the days since i installed the integration)

is it normal that i dont see any other iec reports for october?

![image](https://github.com/user-attachments/assets/8cb1cf9b-61d4-45a3-bca7-8311cf3c2793)

**Reply by GuyKh** [reply]
It's catching up.
Look on a daily resolution and check in an hour and in two.
If all works well, additional days would be added.

**Reply by smarg120** [reply]
i am looking on a daily resolution and dont see anything there....going back day by day - i dont see any new data arriving....going back to september 12 - then suddenly i see the report for the whole month 

how can i check if its indeed running  and adding something? anything specific in logs

**Reply by smarg120** [reply]
![image](https://github.com/user-attachments/assets/8df867ad-d5c7-46b1-95a6-160e2c050c8b)

**Reply by GuyKh** [reply]
Set the logs to `debug` and check for `iec` in the logs.
See more instructions in README

**Reply by smarg120** [reply]
ok, so i enabled the debug logs, thats what i see...every hour there are 3 requests

first request -  HTTP POST: https://iecapi.iec.co.il/api/Consumption/RemoteReadingRange/xxxxx
with exactly the same body - b'{"meterSerialNumber":"xxxx","meterCode":"xxxx","lastInvoiceDate":"2024-09-12","fromDate":"2024-09-12","resolution":1}'
and it gets exactly the same response....and it keeps asking again for the same "fromDate":"2024-09-12" and gets statistics for that day, over and over again

second request - HTTP POST: https://iecapi.iec.co.il/api/Consumption/RemoteReadingRange/xxxxx with body
b'{"meterSerialNumber":"xxxxx","meterCode":"xxxx","lastInvoiceDate":"2024-10-16","fromDate":"2024-10-16","resolution":3}'
the response contains perfect and correct data for the current month, with correct consumption and dates! 
but for whatever reason it does not appear in statistics graph


last request has the body of b'{"meterSerialNumber":"xxx","meterCode":"xxx","lastInvoiceDate":"2024-10-15","fromDate":"2024-10-15","resolution":1}'  and this brings just the data for today, which is off course not accurate yet as was not updated by IEC


so to summarize i have 2 issues
1) less important but i am just curios why the first request keeps asking for the same historic data of 12/9....looks like its stuck on that date
2) this is my major problem - and it maybe related to first issue - although the second request brings correct consumption for the dates, it does not appear at all in the dashboard

thank you for your time, help and responses, i would really love to make it work

**Reply by smarg120** [reply]
also every hour i keep seeing this:

2024-10-15 16:53:16.707 DEBUG (MainThread) [custom_components.iec.coordinator] Last statistics are from 2024-09-12 20:00:00
2024-10-15 16:53:16.707 DEBUG (MainThread) [custom_components.iec.coordinator] Fetching consumption from 2024-09-12 20:00:00

**Reply by smarg120** [reply]
i thought maybe its some db corruption.....so i deleted LTS for those iec meters from the database...and now it does the same loop but for 17/9....

024-10-15 20:12:17.409 DEBUG (MainThread) [custom_components.iec.coordinator] Last statistics are from 2024-09-17 20:00:00
2024-10-15 20:12:17.409 DEBUG (MainThread) [custom_components.iec.coordinator] Fetching consumption from 2024-09-17 20:00:00

**Reply by smarg120** [reply]
it stuck on this hour and it keeps reporting the same data for that hour again and again, so that bar at 9PM just keeps growing...

![image](https://github.com/user-attachments/assets/5913c1b2-3520-42bf-95f3-dfd16abe0618)

**Reply by smarg120** [reply, marked as answer]
ok, I figured it out finally. my ha box had TZ configured to utc. once I configured it to  IL timezone the integration started to record statistics for new days!

thanks for your patience

**Reply by GuyKh** [reply]
Interesting.

Being "stuck" on that date is since it tries to fetch from that time and this is the last hour it has information about.

I'll try to look into why the TZ causes this loop. Thank you for your patience and sharing the data.

**Reply by smarg120** [reply]
Yeah, now it populates perfectly. 
I didnt had patience to wait  for it to catchup, just reloaded the integration multiple times and last month is populated. 

Now I can monitor my consumption usage, thank you!
